lib/tesla/adapter/mint.ex

```diff
@@ -0,0 +1,184 @@
+      for {key, value} <- headers do
+        {String.downcase(to_string(key)), to_string(value)}
+      end
+    end
+
+    defp request(env, opts) do
+      # Break the URI
+      %URI{host: host, scheme: scheme, port: port, path: path, query: query} = URI.parse(env.url)
+      query = (query || "") |> URI.decode_query() |> Map.to_list()
+      path = Tesla.build_url(path, env.query ++ query)
+
+      method =
+        case env.method do
+          :head -> "GET"
+          m -> m |> Atom.to_string() |> String.upcase()
+        end
+
+      opts =
+        if opts |> get_in([:transport_opts, :cacertfile]) |> is_nil() && scheme == "https" &&
+             !is_nil(get_default_ca()) do
+          transport =
+            opts
+            |> Access.get(:transport_opts, [])
+            |> update_in([:cacertfile], fn _ ->
+              get_default_ca()
+            end)
+
+          update_in(opts, [:transport_opts], fn _ ->
+            transport
+          end)
+        else
+          opts
+        end
+
+      request(
+        method,
+        scheme,
+        host,
+        port,
+        path,
+        env.headers,
+        env.body,
+        opts
+      )
+    end
+
+    defp request(method, scheme, host, port, path, headers, %Stream{} = body, opts) do
+      fun = stream_to_fun(body)
+      request(method, scheme, host, port, path, headers, fun, opts)
+    end
+
+    defp request(method, scheme, host, port, path, headers, %Multipart{} = body, opts) do
+      headers = headers ++ Multipart.headers(body)
+      fun = stream_to_fun(Multipart.body(body))
+      request(method, scheme, host, port, path, headers, fun, opts)
+    end
+
+    defp request(method, scheme, host, port, path, headers, body, opts) when is_function(body) do
+      with {:ok, conn} <- HTTP.connect(String.to_atom(scheme), host, port, opts),
+           # FIXME Stream function in Mint will not append the content length after eof
+           # This will trigger the failure in unit test
+           {:ok, body, length} <- stream_request(body),
+           {:ok, conn, _req_ref} <-
+             HTTP.request(
+               conn,
+               method,
+               path || "/",
+               headers ++ [{"content-length", "#{length}"}],
+               body
+             ),
+           {:ok, _conn, res = %{status: status, headers: headers}} <- stream_response(conn) do
+        {:ok, status, headers, Map.get(res, :data)}
+      end
+    end
+
+    defp request(method, scheme, host, port, path, headers, body, opts) do
+      with {:ok, conn} <- HTTP.connect(String.to_atom(scheme), host, port, opts),
+           {:ok, conn, _req_ref} <- HTTP.request(conn, method, path || "/", headers, body || ""),
+           {:ok, _conn, res = %{status: status, headers: headers}} <- stream_response(conn) do
+        {:ok, status, headers, Map.get(res, :data)}
+      end
+    end
+
+    defp get_default_ca() do
+      Application.get_env(:tesla, :cacert)
+    end
```

Comment:
This should be scoped to mint adapter to prevent confusion when using other adapters